### PR TITLE
xjalienfs: always relocate python

### PR DESCRIPTION
**Pull request by ktf**

### CODE REVIEW

**Comment by ktf** [comment]
@aphecetche any reason not to always relocate?

**Comment by aphecetche** [comment]
@ktf trying this PR with my setup I get : 

```
[ (python) brew ]~/alice/brew$ cat /Users/laurent/alice/brew/sw/BUILD/xjalienfs-latest/log
++ PIPOPTION=--user
++ '[' '!' X/Users/laurent/alice/brew/.venv/brew = X ']'
++ PIPOPTION=
++ env PYTHONUSERBASE=/Users/laurent/alice/brew/sw/INSTALLROOT/3c31027e4f6fa653d613b4be0df12081072fdf6e/osx_x86-64/xjalienfs/1.2.8-4 ALIBUILD=1 python3 -m pip install --ignore-installed file:///Users/laurent/alice/brew/sw/SOURCES/xjalienfs/1.2.8/1.2.8
Processing /Users/laurent/alice/brew/sw/SOURCES/xjalienfs/1.2.8/1.2.8
Collecting async-stagger
  Using cached async_stagger-0.3.1-py3-none-any.whl (25 kB)
Collecting websockets
  Using cached websockets-8.1-cp39-cp39-macosx_11_0_x86_64.whl
Collecting pyOpenSSL
  Using cached pyOpenSSL-20.0.1-py2.py3-none-any.whl (54 kB)
Collecting gnureadline
  Using cached gnureadline-8.0.0-cp39-cp39-macosx_11_0_x86_64.whl
Collecting cryptography>=3.2
  Using cached cryptography-3.4.6-cp36-abi3-macosx_10_10_x86_64.whl (2.0 MB)
Collecting six>=1.5.2
  Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting cffi>=1.12
  Using cached cffi-1.14.5-cp39-cp39-macosx_10_9_x86_64.whl (177 kB)
Collecting pycparser
  Using cached pycparser-2.20-py2.py3-none-any.whl (112 kB)
Building wheels for collected packages: alienpy
  Building wheel for alienpy (setup.py): started
  Building wheel for alienpy (setup.py): finished with status 'done'
  Created wheel for alienpy: filename=alienpy-1.2.8-py3-none-any.whl size=45754 sha256=d3c0aab93b2d8b3e2e4be558060d6660f2e4104cfac77f7088d5e3f3594d898c
  Stored in directory: /private/var/folders/sk/tqly2lp512s0n9c1pc2njkkc0000gn/T/pip-ephem-wheel-cache-ukzd2x_f/wheels/1b/34/36/60cca49ecc9879c94728f5f227611bf707ccf71fdffa366026
Successfully built alienpy
Installing collected packages: pycparser, cffi, six, cryptography, websockets, pyOpenSSL, gnureadline, async-stagger, alienpy
Successfully installed alienpy-1.2.8 async-stagger-0.3.1 cffi-1.14.5 cryptography-3.4.6 gnureadline-8.0.0 pyOpenSSL-20.0.1 pycparser-2.20 six-1.15.0 websockets-8.1
++ sed -i.bak 's/#!.*python.*/#!\/usr\/bin\/env python3/' '/Users/laurent/alice/brew/sw/INSTALLROOT/3c31027e4f6fa653d613b4be0df12081072fdf6e/osx_x86-64/xjalienfs/1.2.8-4/bin/*'
sed: /Users/laurent/alice/brew/sw/INSTALLROOT/3c31027e4f6fa653d613b4be0df12081072fdf6e/osx_x86-64/xjalienfs/1.2.8-4/bin/*: No such file or directory
```

"My setup" being that I do all my Alice related work from within a pre-existing python virtualenv.

```
echo $VIRTUAL_ENV
/Users/laurent/alice/brew/.venv/brew
```

**Comment by ktf** [comment]
Hopefully fixed in a better way...

On March 18, 2021, ALICE Builder ***@***.***> wrote:
> @ktf <https://github.com/ktf> trying this PR with my setup I get :
>
> [ (python) brew ]~/alice/brew$ cat
> /Users/laurent/alice/brew/sw/BUILD/xjalienfs-latest/log ++
> PIPOPTION=--user ++ '[' '!' X/Users/laurent/alice/brew/.venv/brew = X
> ']' ++ PIPOPTION= ++ env
> PYTHONUSERBASE=/Users/laurent/alice/brew/sw/INSTALLROOT/3c31027e4f6fa653d613b4be0df12081072fdf6e/osx_x86-
> 64/xjalienfs/1.2.8-4 ALIBUILD=1 python3 -m pip install --ignore-
> installed
> file:///Users/laurent/alice/brew/sw/SOURCES/xjalienfs/1.2.8/1.2.8
> Processing /Users/laurent/alice/brew/sw/SOURCES/xjalienfs/1.2.8/1.2.8
> Collecting async-stagger Using cached async_stagger-0.3.1-py3-none-
> any.whl (25 kB) Collecting websockets Using cached websockets-8.1-
> cp39-cp39-macosx_11_0_x86_64.whl Collecting pyOpenSSL Using cached
> pyOpenSSL-20.0.1-py2.py3-none-any.whl (54 kB) Collecting gnureadline
> Using cached gnureadline-8.0.0-cp39-cp39-macosx_11_0_x86_64.whl
> Collecting cryptography>=3.2 Using cached cryptography-3.4.6-cp36-
> abi3-macosx_10_10_x86_64.whl (2.0 MB) Collecting six>=1.5.2 Using
> cached six-1.15.0-py2.py3-none-any.whl (10 kB) Collecting cffi>=1.12
> Using cached cffi-1.14.5-cp39-cp39-macosx_10_9_x86_64.whl (177 kB)
> Collecting pycparser Using cached pycparser-2.20-py2.py3-none-any.whl
> (112 kB) Building wheels for collected packages: alienpy Building
> wheel for alienpy (setup.py): started Building wheel for alienpy
> (setup.py): finished with status 'done' Created wheel for alienpy:
> filename=alienpy-1.2.8-py3-none-any.whl size=45754
> sha256=d3c0aab93b2d8b3e2e4be558060d6660f2e4104cfac77f7088d5e3f3594d898c
> Stored in directory:
> /private/var/folders/sk/tqly2lp512s0n9c1pc2njkkc0000gn/T/pip-ephem-
> wheel-cache-
> ukzd2x_f/wheels/1b/34/36/60cca49ecc9879c94728f5f227611bf707ccf71fdffa366026
> Successfully built alienpy Installing collected packages: pycparser,
> cffi, six, cryptography, websockets, pyOpenSSL, gnureadline, async-
> stagger, alienpy Successfully installed alienpy-1.2.8 async-stagger-
> 0.3.1 cffi-1.14.5 cryptography-3.4.6 gnureadline-8.0.0 pyOpenSSL-
> 20.0.1 pycparser-2.20 six-1.15.0 websockets-8.1 ++ sed -i.bak
> 's/#!.*python.*/#!\/usr\/bin\/env python3/'
> '/Users/laurent/alice/brew/sw/INSTALLROOT/3c31027e4f6fa653d613b4be0df12081072fdf6e/osx_x86-
> 64/xjalienfs/1.2.8-4/bin/*' sed:
> /Users/laurent/alice/brew/sw/INSTALLROOT/3c31027e4f6fa653d613b4be0df12081072fdf6e/osx_x86-
> 64/xjalienfs/1.2.8-4/bin/*: No such file or directory
> "My setup" being that I do all my Alice related work from within a
> pre-existing python virtualenv.
>
> echo $VIRTUAL_ENV /Users/laurent/alice/brew/.venv/brew

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/alisw/alidist/pull/2911#issuecomment-801684932>, or
unsubscribe <https://github.com/notifications/unsubscribe-
auth/AAACSMDWEIZ5XVZTP3YSEMLTEGQ3FANCNFSM4ZLK4EBQ>.

**Comment by aphecetche** [comment]
@ktf yep, now `aliBuild --defaults o2 build xjalienfs` ends up happy.

File: xjalienfs.sh [hunk 1]
```diff
@@ -18,15 +18,14 @@ fi
...
-   # Uniform Python library path
-  pushd ${INSTALLROOT}
-    pushd lib
+# Make sure all the tools use the correct python
+mkdir -p ${INSTALLROOT}/bin 
+find ${INSTALLROOT}/bin -type f -exec sed -i".bak" 's/#!.*python.*/#!\/usr\/bin\/env python3/' \;
+rm -fv ${INSTALLROOT}/bin/*.bak
+
+if [ -d ${INSTALLROOT}/lib ]; then
+  pushd ${INSTALLROOT}/lib
```
Comment (sawenzel) [inline xjalienfs.sh:27]: just for my understanding: Why is not enough to do consistently `#!/usr/bin/env python3` in our python scripts? @adriansev @ktf

Comment (ktf) [reply xjalienfs.sh:27]: The problem is that when installing with pip, it uses the path of the currently used python.

Comment (adriansev) [reply xjalienfs.sh:27]: @sawenzel if you are asking why is the sed line is required for `#!/usr/bin/env python3` the reason is that the `entry_points console_scripts` mechanism of setuptools (see setup.py) place that less than optimal shebang (so outside on automatic control) .. i seen some discussions on the topic but i'm not aware of any resolution on that.

**Comment by adriansev** [comment]
@ktf @sawenzel hi guys! it was noticed this:
```
diff -u /cvmfs/alice.cern.ch/el7-x86_64/Packages/xjalienfs/1.2.8-1/bin/alien.py /cvmfs/alice.cern.ch/el7-x86_64/Packages/xjalienfs/1.2.9-1/bin/alien.py
--- /cvmfs/alice.cern.ch/el7-x86_64/Packages/xjalienfs/1.2.8-1/bin/alien.py     2020-11-24 04:53:00.000000000 +0200
+++ /cvmfs/alice.cern.ch/el7-x86_64/Packages/xjalienfs/1.2.9-1/bin/alien.py     2021-03-20 04:46:43.000000000 +0200
@@ -1,11 +1,10 @@
-#!/usr/bin/env python3
-
+#!/bin/sh
+'''exec' /mnt/mesos/sandbox/sandbox/jenkins/workspace/DailyBuilds/DailyO2/daily-tags.Y7p0Rlmjnm/slc7_x86-64/Python/v3.6.10-109/bin/python3 "$0" "$@"
+' '''
# -*- coding: utf-8 -*-
import re
import sys
-
from alienpy.alien import main
-
if __name__ == '__main__':
-    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
+    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
sys.exit(main())
```
and as such 
```
/cvmfs/alice.cern.ch/el7-x86_64/Packages/xjalienfs/1.2.9-1/bin/alien.py: 2: exec: /mnt/mesos/sandbox/sandbox/jenkins/workspace/DailyBuilds/DailyO2/daily-tags.Y7p0Rlmjnm/slc7_x86-64/Python/v3.6.10-109/bin/python3: not found
```
I'm not sure where is the problem but maybe is related to this commit?

**Pull request merged**